File: backend/app/schemas/creator_intelligence.py
"""Creator intelligence dashboard — growth, hooks, audience, semantic, momentum."""

from datetime import date

from pydantic import BaseModel, Field

from app.schemas.analytics import ChartPoint, HookAnalysis, KeywordStat
from app.schemas.comments import CommentRead
from app.schemas.creator_page import CreatorOverview, CreatorPageSections
from app.schemas.growth import VideoBreakoutItem
from app.schemas.video import VideoRead


class CreatorGrowthMetrics(BaseModel):
    """Point-in-time growth from snapshots."""

    growth_7d_pct: float = 0.0
    growth_30d_pct: float = 0.0
    subscribers_delta_7d: int = 0
    views_delta_7d: int = 0
    velocity_views_per_day: float = 0.0
    snapshot_days: int = 0
    accelerating: bool = False
    slowing: bool = False


class CreatorGrowthIntel(BaseModel):
    metrics: CreatorGrowthMetrics = Field(default_factory=CreatorGrowthMetrics)
    subscriber_history: list[ChartPoint] = Field(default_factory=list)
    views_history: list[ChartPoint] = Field(default_factory=list)
    upload_momentum: list[ChartPoint] = Field(default_factory=list)
    latest_snapshot_date: date | None = None


class CreatorHookMix(BaseModel):
    """Title hook mix percentages for one creator."""

    curiosity_pct: float = 0.0
    transformation_pct: float = 0.0
    urgency_pct: float = 0.0
    numbers_pct: float = 0.0
    emotional_pct: float = 0.0
    authority_pct: float = 0.0
    how_to_pct: float = 0.0
    identity_pct: float = 0.0


class CreatorHookIntel(BaseModel):
    mix: CreatorHookMix = Field(default_factory=CreatorHookMix)
    analysis: HookAnalysis = Field(default_factory=HookAnalysis)
    best_performing_hooks: list[str] = Field(default_factory=list)


class CreatorAudienceIntel(BaseModel):
    total_comments: int = 0
    top_comments: list[CommentRead] = Field(default_factory=list)
    repeated_phrases: list[str] = Field(default_factory=list)
    emotional_patterns: list[str] = Field(default_factory=list)
    pain_points: list[str] = Field(default_factory=list)
    top_reactions: list[str] = Field(default_factory=list)


class NearestCreator(BaseModel):
    creator_name: str
    overlap_score: float = 0.0
    shared_topics: list[str] = Field(default_factory=list)


class CreatorSemanticIntel(BaseModel):
    dominant_keywords: list[KeywordStat] = Field(default_factory=list)
    themes: list[str] = Field(default_factory=list)
    positioning_summary: str = ""
    nearest_creators: list[NearestCreator] = Field(default_factory=list)


class CreatorMomentumIntel(BaseModel):
    breakout_videos: list[VideoBreakoutItem] = Field(default_factory=list)
    fastest_growing: list[VideoBreakoutItem] = Field(default_factory=list)
    latest_uploads: list[VideoRead] = Field(default_factory=list)


class CreatorIntelligence(BaseModel):
    """Full GET /creators/{name}/intelligence payload."""

    overview: CreatorOverview
    growth: CreatorGrowthIntel = Field(default_factory=CreatorGrowthIntel)
    hooks: CreatorHookIntel = Field(default_factory=CreatorHookIntel)
    audience: CreatorAudienceIntel = Field(default_factory=CreatorAudienceIntel)
    semantic: CreatorSemanticIntel = Field(default_factory=CreatorSemanticIntel)
    momentum: CreatorMomentumIntel = Field(default_factory=CreatorMomentumIntel)
    sections: CreatorPageSections = Field(default_factory=CreatorPageSections)

File: frontend/types/creator-intelligence.ts
/** Creator intelligence dashboard — mirrors backend /creators/{name}/intelligence */

import type { HookAnalysis } from "@/types/analytics";
import type { CommentRead } from "@/types/video-intelligence";
import type { ChartPoint, CreatorOverview, CreatorPageSections } from "@/types/creator-page";
import type { Video } from "@/types/video";

export interface CreatorGrowthMetrics {
  growth_7d_pct: number;
  growth_30d_pct: number;
  subscribers_delta_7d: number;
  views_delta_7d: number;
  velocity_views_per_day: number;
  snapshot_days: number;
  accelerating: boolean;
  slowing: boolean;
}

export interface CreatorGrowthIntel {
  metrics: CreatorGrowthMetrics;
  subscriber_history: ChartPoint[];
  views_history: ChartPoint[];
  upload_momentum: ChartPoint[];
  latest_snapshot_date: string | null;
}

export interface CreatorHookMix {
  curiosity_pct: number;
  transformation_pct: number;
  urgency_pct: number;
  numbers_pct: number;
  emotional_pct: number;
  authority_pct: number;
  how_to_pct: number;
  identity_pct: number;
}

export interface CreatorHookIntel {
  mix: CreatorHookMix;
  analysis: HookAnalysis;
  best_performing_hooks: string[];
}

export interface CreatorAudienceIntel {
  total_comments: number;
  top_comments: CommentRead[];
  repeated_phrases: string[];
  emotional_patterns: string[];
  pain_points: string[];
  top_reactions: string[];
}

export interface NearestCreator {
  creator_name: string;
  overlap_score: number;
  shared_topics: string[];
}

export interface CreatorSemanticIntel {
  dominant_keywords: { keyword: string; count: number; avg_views: number }[];
  themes: string[];
  positioning_summary: string;
  nearest_creators: NearestCreator[];
}

export interface VideoBreakoutItem {
  video_id: number;
  title: string;
  creator_name: string;
  views_now: number;
  views_delta_7d: number;
  growth_7d_pct: number;
  velocity_views_per_day: number;
  breakout_score: number;
}

export interface CreatorMomentumIntel {
  breakout_videos: VideoBreakoutItem[];
  fastest_growing: VideoBreakoutItem[];
  latest_uploads: Video[];
}

export interface CreatorIntelligence {
  overview: CreatorOverview;
  growth: CreatorGrowthIntel;
  hooks: CreatorHookIntel;
  audience: CreatorAudienceIntel;
  semantic: CreatorSemanticIntel;
  momentum: CreatorMomentumIntel;
  sections: CreatorPageSections;
}
